eth/downloader/skeleton.go

```diff
@@ -1120,26 +1120,41 @@ func (s *skeleton) cleanStales(filled *types.Header) error {
 	number := filled.Number.Uint64()
 	log.Trace("Cleaning stale beacon headers", "filled", number, "hash", filled.Hash())
 
-	// If the filled header is below the linked subchain, something's
-	// corrupted internally. Report and error and refuse to do anything.
-	if number < s.progress.Subchains[0].Tail {
+	// If the filled header is below and discontinuous with the linked subchain,
+	// something's corrupted internally. Report and error and refuse to do anything.
+	if number+1 < s.progress.Subchains[0].Tail {
```

Comment:
I'm unsure that this is correct.

Number is the last item **newly** filled. When I start a sync loop, lets say I have N blocks in my chain. Then I skeleton sync, and will reconstruct headers from HEAD to N+1, and which point I have my subchain linked to the local chain.

I start backfilling. If I fail to fill anything, `cleanStales` does not get called. If I successfully fill 1 block, then `filled` will be N+1 (the newly filled one block).

In that case `number` == `filled` == `N+1`, which is == `s.progress.Subchains[0].Tail`. The only way to hit this error would be to backfill something we already had in our chain. 

With the proposed modification however, backfilling my local head block would make the check pass, but that should not be possible, because I can only backfill stuff from my subchain, not 1 block below it.

Reply:
According the below, the will only call cleanStales if *something* was filled.

```go
			if filled := s.filler.suspend(); filled != nil {
				// If something was filled, try to delete stale sync helpers. If
				// unsuccessful, warn the user, but not much else we can do (it's
				// a programming error, just let users report an issue and don't
				// choke in the meantime).
				if err := s.cleanStales(filled); err != nil {
					log.Error("Failed to clean stale beacon headers", "err", err)
				}
			}
```

That said, you are also right that resuming the filler will always return the head snap block (or genesis I guess in case we're brand new).

```go
		defer func() {
			b.lock.Lock()
			b.filling = false
			b.filled = b.downloader.blockchain.CurrentSnapBlock()
			b.lock.Unlock()
		}()
```

Interesting, I guess this was a refactor after I designed the original idea. It kind of makes the filled check in my first code segment moot.